PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionJobUploadStoragePanel.kt

```diff
@@ -174,24 +140,97 @@ open class SparkSubmissionJobUploadStoragePanel
+            assertInDispatchThread()
+
+            if (clusterDetail == null) {
+                return deployStorageTypeSelection
+            }
+
+            val optionTypes = clusterDetail.storageOptionsType.optionTypes
+            val currentStorageTypesModel = deployStorageTypesModel
+            val newStorageTypesModel = ImmutableComboBoxModel(optionTypes)
+
+            val storageTypesModelToSet = if (currentStorageTypesModel.size == optionTypes.size) {
+                // Deep compare the size and items
+                var isDeepEqualed = true
+                for (i in 0 until currentStorageTypesModel.size) {
+                    if (currentStorageTypesModel.getElementAt(i) != optionTypes[i]) {
+                        isDeepEqualed = false
+                        break
+                    }
+                }
+
+                if (isDeepEqualed) currentStorageTypesModel else newStorageTypesModel
+            } else newStorageTypesModel
+
+            if (storageTypesModelToSet != currentStorageTypesModel) {
+                // there exist 4 cases to set the storage type
+                // 1.select cluster -> set to default
+                // 2.reload config with not null type -> set to saved type
+                // 3.reload config with null storage type -> set to default
+                // 4.create config  -> set to default
+                deployStorageTypesModel = storageTypesModelToSet
+                deployStorageTypeSelection = clusterDetail.defaultStorageType
+            }
+
+            return storageTypesModelToSet.selectedItem as? SparkSubmitStorageType ?: clusterDetail.defaultStorageType
+        }
+
+        val errorMessage: String?
+            get() {
+                val currentCardViewModel = this.currentCardViewModel
+                        ?: return "No selected artifacts uploading storage type"
+
+                return currentCardViewModel.errorMessage
+            }
+
+        init {
+            // Merge all storage cards storage validated upload URI events into the panel's
+            Observable.merge(storageCards.values.map { it.viewModel.validatedStorageUploadUri })
+                      .subscribe(validatedStorageUploadUri::onNext,
+                                 validatedStorageUploadUri::onError,
+                                 validatedStorageUploadUri::onCompleted)
+
+            clusterSelectedSubject
+                    .doOnNext {
+                        runInEdt(stateForComponent(view)) {
+                            log().info("Current model storage account type is $deployStorageTypeSelection")
+                            updateStorageTypesModelForCluster(it)
+                            log().info("Update model storage account type to $deployStorageTypeSelection")
+
+                            currentCardViewModel?.cluster = it
+                        }
+                    }
+                    .subscribe()
```

Comment:
Better to handle errors when we `subscribe` a `Subject`

Reply:
Good catch! Thanks!